client/hotrod-client/src/main/java/org/infinispan/client/hotrod/impl/iteration/SegmentKeyTracker.java

```diff
@@ -85,7 +85,11 @@ public void segmentsFinished(IntSet finishedSegments) {
       if (trackSegments && finishedSegments != null) {
          if (log.isTraceEnabled())
             log.tracef("Removing completed segments %s", finishedSegments);
-         finishedSegments.forEach((IntConsumer) seg -> keysPerSegment.set(seg, null));
+         finishedSegments.forEach((IntConsumer) seg -> {
+            Set<WrappedByteArray> keys = keysPerSegment.get(seg);
+            if (keys != null) keyOnlyTracker.addAll(keys);
```

Comment:
This looks very wrong to me. Are you sure this fixes anything? As far as I can see the `keysOnlyTracker` should only be used if `trackSegments` is false. This screams as a memory issue to me as every single key that was tracked will be stored in this Set at one time, causing all sorts of memory issues.

Reply:
TBH this whole `keysOnlyTracker` thing seems like it is trying to cover up a bug somewhere else, as we shouldn't ever receive a key for a segment that has since been completed. I would say we should throw an exception instead when we receive a key that maps to an already completed segment and not even have this `keysOnlyTracker` set as it is an OOM issue waiting to explode.

Reply:
IMO we should go with throwing an exception here instead of the `keysOnlyTracker`. I changed this because, originally, a key could be accepted twice. Now with the changes for the IntSet, we shouldn't receive keys for a segment already completed.

Reply:
My thought is to remove `keysOnlyTracker` completely and in the `track` method if we receive a key that doesn't map to a segment then throw the exception.